File: java/src/main/java/org/sammancoaching/DeployStep.java
package org.sammancoaching;

import org.sammancoaching.dependencies.Config;
import org.sammancoaching.dependencies.Logger;
import org.sammancoaching.dependencies.Project;

public record DeployStep(String name, Config config, Logger log) implements PipelineStep {
    public PipelineStepResult run(Project project, PipelineStepResult previousStepResult) {
        boolean stepPassed = false;
        String failureReason = previousStepResult.failureReason();
        if (previousStepResult.stepPassed()) {
            if ("success".equals(project.deploy())) {
                log.info(name() + " successful");
                stepPassed = true;
            } else {
                String reason = name() + " failed";
                log.error(reason);
                stepPassed = false;
                failureReason = reason;
            }
        }
        return new PipelineStepResult(this.name(), stepPassed, failureReason);
    }
}

File: csharp/BuildPipeline/DeployStep.cs
namespace UntangledConditionals
{
    public class DeployStep : PipelineStep
    {
        private readonly string _name;
        private readonly Config _config;
        private readonly Logger _log;

        public DeployStep(string name, Config config, Logger log)
        {
            _name = name;
            _config = config;
            _log = log;
        }

        public PipelineStepResult Run(Project project, PipelineStepResult previousStepResult)
        {
            bool stepPassed = false;
            string failureReason = previousStepResult.FailureReason;

            if (previousStepResult.StepPassed)
            {
                if ("success".Equals(project.Deploy()))
                {
                    _log.Info(_name + " successful");
                    stepPassed = true;
                }
                else
                {
                    string reason = _name + " failed";
                    _log.Error(reason);
                    stepPassed = false;
                    failureReason = reason;
                }
            }

            return new PipelineStepResult(_name, stepPassed, failureReason);
        }
    }
}

